### PR TITLE
FIX-2103 Compare log data unequal server decimals

### LINKED ISSUE #2103: 💡Support depth log data comparison on servers with different numbers of decimals for index
**💡 Is your feature request related to a problem? Please describe.**
Currently, comparing log data is only supported on servers with equal numbers of decimals for the depth log indexes. Supporting cases where two servers have different numbers of decimals is necessary.

**Describe the solution you'd like**
To compare log data, follow these steps: right-click on a depth log and select "Compare log" and then click on "Compare log data". You can then compare the selected log with a log on a server supporting a different number of decimals than the chosen log.

Some of the code in [CopyLogDataWorker.cs](https://github.com/equinor/witsml-explorer/blob/b144d33359c667327def27a0ce5badf1885a35ab/Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs#L189) can be reused.

**Example:**

<table>
<tr><th>Log 1</th><th>Log 2</th></tr>
<tr><td>

| index | data |
|-------|------|
| 0.01  | 11   |
| 0.02  | 22   |
| 0.03  | 33   |
| 0.04  | 44   |

</td><td>

| index | data |
|-------|------|
| 0.012 | 11   |
| 0.013 | 88   |
| 0.014 | 11   |
| 0.018 | 22   |
| 0.026 | 99   |
| 0.038 | 44   |

</td></tr> </table>

Find the log with the most decimal places. In this case, it is "Log 2". Then round the indexes to the same number of decimal places as "Log 1":

<table>
<tr><th>Log 1</th><th>Log 2 rounded</th></tr>
<tr><td>

| index | data |
|-------|------|
| 0.01  | 11   |
| 0.02  | 22   |
| 0.03  | 33   |
| 0.04  | 44   |

</td><td>

| index | old_index | data |
|-------|-----------|------|
| 0.01  | 0.012     | 11   |
| 0.01  | 0.013     | 88   |
| 0.01 | 0.014      | 11   |
| 0.02  | 0.018     | 22   |
| 0.03  | 0.026     | 99   |
| 0.04  | 0.038     | 44   |

</td></tr> </table>

When comparing the data with indexes that have an equal number of decimals, the resulting output should be:
| mnemonic                | sourceIndex | targetIndex | sourceValue | targetValue |
|-------------------------|-------------|-------------|-------------|-------------|
| not included in example | 0.01        | 0.013       | 11          | 88          |
| not included in example | 0.03        | 0.026       | 33         | 99          |

### PR DESCRIPTION
**Pull request by janmarius**
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes

[//]: # (Write the GitHub issue number starting with #, or, for Equinor only, the Jira issue number starting with WE-)

This pull request fixes #2103 

## Description

[//]: # (Please include a written summary of the changes and which issue is fixed. List any dependencies that are required for this change._)

This pull request adds support to `CompareLogDataWorker` for depth log indexes with unequal decimals on servers. Additional tests have been included.

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [ ] New feature (non-breaking change which adds functionality)
* [x] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [ ] Frontend
* [x] API
* [ ] WITSML
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [x] I have self-reviewed my code
* [x] No new warnings are generated

*Test coverage*
* [x] New code is covered by passing tests

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

The expected result in the issue example is incorrect. We want to report all rounded indexes that are duplicated, not only those that mismatch.

### CODE REVIEW

File: Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs [hunk 1]
```diff
@@ -38,16 +43,10 @@ public override async Task<(WorkerResult, RefreshAction)> Execute(CompareLogData
...
-            // Check if the number of decimals for both servers is equal. Throw an error if not.
-            if (sourceDepthLogDecimals != targetDepthLogDecimals)
-            {
-                string message = $"CompareLogDataJob failed. Cases where servers have different numbers of decimals are not currently supported.";
-                Logger.LogError(message);
-                return (new WorkerResult(GetSourceWitsmlClientOrThrow().GetServerHostname(), false, message), null);
-            }
+            _sourceDepthLogDecimals = servers.FirstOrDefault((server) => server.Url.EqualsIgnoreCase(sourceHostname))?.DepthLogDecimals ?? 0;
+            _targetDepthLogDecimals = servers.FirstOrDefault((server) => server.Url.EqualsIgnoreCase(targetHostname))?.DepthLogDecimals ?? 0;
+            _smallestDepthLogDecimals = _sourceDepthLogDecimals <= _targetDepthLogDecimals ? _sourceDepthLogDecimals : _targetDepthLogDecimals;
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs]: Looks a bit cleaner to use buildt-in functions such as 
`_smallestDepthLogDecimals = Math.Min(_sourceDepthLogDecimals, _targetDepthLogDecimals);`

File: Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs [hunk 2]
```diff
@@ -151,33 +207,55 @@ private async Task AddSharedMnemonicData(WitsmlLog sourceLog, WitsmlLog targetLo
...
+                else if (moreDecimalsIndexes.Contains(index) && !lessDecimalsIndexes.Contains(index))
                 {
-                    throw new ArgumentException($"serverType={serverType} not supported.");
+                    string moreDecimalsValue = Logs.MoreDecimals[index];
+                    AddUnequalServerDecimalsReportItem(mnemonic, null, index, null, moreDecimalsValue);
                 }
             }
         }
 
+        private bool IsMnemonicDataEquals(string sourceValue, string targetValue)
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs]: Remove plural "s" from Equals

**Review by eliasbruvik — CHANGES_REQUESTED** [review]
Overall looks good, not much to add.

File: Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs (hunk 1 above)
Comment (janmarius) [reply Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs]: Nice, this looks much cleaner! Fixed.

File: Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs (hunk 2 above)
Comment (janmarius) [reply Src/WitsmlExplorer.Api/Workers/CompareLogDataWorker.cs]: Good catch! Fixed.

**Review by eliasbruvik — APPROVED** [review]
- [x] Read issue and PR
- [x] Pulled branch and manually tested
- [x] Verify that PR resolves issue
- [x] Reviewed the code

**Pull request merged**